nx/lib/nx/lin_alg.ex

```diff
@@ -967,4 +967,88 @@ defmodule Nx.LinAlg do
       opts
     )
   end
+
+  @doc """
+  Produces the tensor taken to the given power by dot-product.
+
+  The input is always a square tensor and a non-negative integer,
+  and the output is a square tensor of the same dimensions as the input tensor.
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 6)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [5743, 8370],
+          [12555, 18298]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.eye(3), 65535)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), -1)
+      #Nx.Tensor<
+        f32[2][2]
+        [
+          [-2.0, 1.0],
+          [1.5, -0.5]
+        ]
+      >
+  """
+  @doc from_backend: false
+  def matrix_power(matrix, power) when is_integer(power) and power < 0 do
+    matrix_power(Nx.LinAlg.invert(matrix), -power)
+  end
+
+  def matrix_power(matrix, 0) do
+    # We need a special-case for 0 because Integer.digits(0, 2) returns
+    # [0]. The last function clause assumes:
+    #     [1 | _] = Enum.reverse(Integer.digits(???, 2))
+    # which is the case for all positive integers, but not for 0.
+    Nx.eye(matrix)
+  end
+
+  def matrix_power(matrix, power) when is_integer(power) do
+    {result, exp_matrix} =
+      power
+      |> Integer.digits(2)
+      |> tl()
+      |> Enum.reverse()
+      |> Enum.reduce({Nx.eye(matrix), matrix}, fn
+        1, {result_matrix, exp_matrix} ->
+          {Nx.dot(result_matrix, exp_matrix), Nx.dot(exp_matrix, exp_matrix)}
+
+        0, {result_matrix, exp_matrix} ->
+          {result_matrix, Nx.dot(exp_matrix, exp_matrix)}
+      end)
+
+    Nx.dot(result, exp_matrix)
+  end
```

Comment:
If you want to remove the special case for 0:

```suggestion
  def matrix_power(matrix, power) when is_integer(power) do
    power
    |> Integer.digits(2)
    |> Enum.reverse()
    |> Enum.reduce({Nx.eye(matrix), matrix}, fn
      1, {result_matrix, exp_matrix} ->
        {Nx.dot(result_matrix, exp_matrix), Nx.dot(exp_matrix, exp_matrix)}

      0, {result_matrix, exp_matrix} ->
        {result_matrix, Nx.dot(exp_matrix, exp_matrix)}
    end)
    |> elem(0)
  end
```

However, there is a small issue where we are always starting from with an eye multiplication, which we can actually optimize out. So I would keep the special case and then do the other suggestion I will post next.

Reply:
```suggestion
  def matrix_power(matrix, 0) do
    # We need a special-case for 0 since the code below
    # is optimized to not compute an initial eye.
    Nx.eye(matrix)
  end

  def matrix_power(matrix, power) when is_integer(power) do
    power
    |> Integer.digits(2)
    |> Enum.reverse()
    |> Enum.reduce({nil, matrix}, fn
      # It is guaranteed that we will have at least one digit, when that happens,
      # we start with exp_matrix to avoid multiplying by eye.
      1, {nil, exp_matrix} ->
        {exp_matrix, Nx.dot(exp_matrix, exp_matrix)}

      1, {result_matrix, exp_matrix} ->
        {Nx.dot(result_matrix, exp_matrix), Nx.dot(exp_matrix, exp_matrix)}

      0, {result_matrix, exp_matrix} ->
        {result_matrix, Nx.dot(exp_matrix, exp_matrix)}
    end)
    |> elem(0)
  end
```

Reply:
Theoretically we can optimize the eye operation away, but only inside `defn`, so I would ship with the optimized form.

Reply:
I specifically changed from an implementation similar to yours to the special 0-case because the implementation you proposed computes one extra fairly large dot-product. Specifically that one might (under certain circumstances) take about 33% of the full execution time of this function.